cmd/clusterctl/clusterdeployer/clusterclient/clusterclient.go

```diff
@@ -128,14 +128,31 @@ func (c *client) EnsureNamespace(namespaceName string) error {
 		return errors.Wrap(err, "error creating core clientset")
 	}
 
-	namespace := apiv1.Namespace{
-		ObjectMeta: metav1.ObjectMeta{
-			Name: namespaceName,
-		},
-	}
-	_, err = clientset.CoreV1().Namespaces().Create(&namespace)
-	if err != nil && !apierrors.IsAlreadyExists(err) {
-		return err
+	_, err = clientset.CoreV1().Namespaces().Get(namespaceName, metav1.GetOptions{})
```

Comment:
How about restructuring it like this?

```go
// Happy path - return quickly
if err == nil {
  return nil
}

if apierrors.IsForbidden(err) {
  // get was forbidden. see if list works

  namespaces, err := clientset.CoreV1().Namespaces().List(metav1.ListOptions{})
  if apierrors.IsForbidden(err) {
    // either return err, which means we halt work
    // or return nil, and either things will be ok (user has permissions to create CAPI resources but they can't check for namespace existence), or things won't (they can't create CAPI resources, or the namespace doesn't exist)
  }
  if err != nil {
    return err
  }
  // check the list
}

if !apierrors.IsNotFound(err) {
  // err is not Forbidden or NotFound - return it
  return err
}

// try to create
// ...

Reply:
@ncdc I'm sorry for long delay. Thank you so much for the proposal. Could you please take a look again.